tls/s2n_client_extensions.c

```diff
@@ -229,27 +218,74 @@ static int s2n_recv_client_server_name(struct s2n_connection *conn, struct s2n_s
     return 0;
 }
 
-static int s2n_recv_client_signature_algorithms(struct s2n_connection *conn, struct s2n_stuffer *extension)
+int s2n_send_client_signature_algorithms_extension(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    /* The extension header */
+    GUARD(s2n_stuffer_write_uint16(out, TLS_EXTENSION_SIGNATURE_ALGORITHMS));
+
+    /* Each hash-signature-alg pair is two bytes, and there's another two bytes for
+     * the extension length field.
+     */
+    uint16_t preferred_hashes_len = sizeof(s2n_preferred_hashes) / sizeof(s2n_preferred_hashes[0]);
+    uint16_t preferred_hashes_size = preferred_hashes_len * 2;
+    uint16_t extension_len_field_size = 2;
+
+    GUARD(s2n_stuffer_write_uint16(out, extension_len_field_size + preferred_hashes_size));
+    GUARD(s2n_send_client_signature_algorithms(out));
+
+    return 0;
+}
+
+int s2n_send_client_signature_algorithms(struct s2n_stuffer *out)
+{
+    /* The array of hashes and signature algorithms we support */
+    uint16_t preferred_hashes_len = sizeof(s2n_preferred_hashes) / sizeof(s2n_preferred_hashes[0]);
+    uint16_t preferred_hashes_size = preferred_hashes_len * 2;
+    GUARD(s2n_stuffer_write_uint16(out, preferred_hashes_size));
+
+    for (int i =  0; i < preferred_hashes_len; i++) {
+        GUARD(s2n_stuffer_write_uint8(out, s2n_preferred_hashes[i]));
+        GUARD(s2n_stuffer_write_uint8(out, TLS_SIGNATURE_ALGORITHM_RSA));
+    }
+    return 0;
+}
+
+int s2n_recv_client_signature_algorithms(struct s2n_connection *conn, struct s2n_stuffer *in, s2n_hash_algorithm *hash, s2n_signature_algorithm *sig)
 {
-    int chosen = sizeof(s2n_preferred_hashes);
     uint16_t length_of_all_pairs;
-    GUARD(s2n_stuffer_read_uint16(extension, &length_of_all_pairs));
-    if (length_of_all_pairs > s2n_stuffer_data_available(extension)) {
+    GUARD(s2n_stuffer_read_uint16(in, &length_of_all_pairs));
+    if (length_of_all_pairs > s2n_stuffer_data_available(in)) {
         /* Malformed length, ignore the extension */
         return 0;
     }
 
-    if (length_of_all_pairs % 2 || s2n_stuffer_data_available(extension) % 2) {
+    if (length_of_all_pairs % 2 || s2n_stuffer_data_available(in) % 2) {
         /* Pairs occur in two byte lengths. Malformed length, ignore the extension. */
         return 0;
     }
 
-    while (s2n_stuffer_data_available(extension)) {
+    int pairs_available = length_of_all_pairs / 2;
+    GUARD(s2n_choose_preferred_signature_hash_algorithm(in, pairs_available, hash, sig));
+
+    return 0;
+}
+
+int s2n_choose_preferred_signature_hash_algorithm(struct s2n_stuffer *in, int pairs_available,
+        s2n_hash_algorithm *hash_alg_out, s2n_signature_algorithm *signature_alg_out)
+{
+    int found_valid_pair = 0;
+    s2n_hash_algorithm best_hash_alg;
+    s2n_signature_algorithm best_signature_alg;
+
+    int pairs_read = 0;
+    int maximum_hash_alg = sizeof(s2n_preferred_hashes) - 1;
+    while (pairs_read < pairs_available) {
```

Comment:
That's an interesting thought. I'll think on it over the weekend. FWIW we use this short circuit logic for cipher suite matching: https://github.com/awslabs/s2n/blob/master/tls/s2n_cipher_suites.c#L568